src/java/org/apache/cassandra/db/memtable/AbstractAllocatorMemtable.java

```diff
@@ -183,26 +184,40 @@ void scheduleFlush()
     {
         int period = metadata().params.memtableFlushPeriodInMs;
         if (period > 0)
+            scheduleFlush(owner, period);
+    }
+
+    private static void scheduleFlush(Owner owner, int period)
+    {
+        logger.trace("scheduling flush in {} ms", period);
+        WrappedRunnable runnable = new WrappedRunnable()
         {
-            logger.trace("scheduling flush in {} ms", period);
-            WrappedRunnable runnable = new WrappedRunnable()
+            protected void runMayThrow()
             {
-                protected void runMayThrow()
-                {
-                    if (isClean())
-                    {
-                        // if we're still clean, instead of swapping just reschedule a flush for later
-                        scheduleFlush();
-                    }
-                    else
-                    {
-                        // we'll be rescheduled by the constructor of the Memtable.
-                        owner.signalFlushRequired(AbstractAllocatorMemtable.this,
-                                                  ColumnFamilyStore.FlushReason.MEMTABLE_PERIOD_EXPIRED);
-                    }
-                }
-            };
-            ScheduledExecutors.scheduledTasks.schedule(runnable, period, TimeUnit.MILLISECONDS);
+                Memtable current = owner.getCurrentMemtable();
+                if (current instanceof AbstractAllocatorMemtable)
+                    ((AbstractAllocatorMemtable) current).flushIfPeriodExpired();
+            }
+        };
+        ScheduledExecutors.scheduledTasks.schedule(runnable, period, TimeUnit.MILLISECONDS);
+    }
+
+    private void flushIfPeriodExpired()
+    {
+        int period = metadata().params.memtableFlushPeriodInMs;
+        if (period > 0 && (System.nanoTime() - creationNano >= TimeUnit.MILLISECONDS.toNanos(period)))
+        {
+            if (isClean())
+            {
+                // if we're still clean, instead of swapping just reschedule a flush for later
+                scheduleFlush(owner, period);
+            }
+            else
+            {
+                // we'll be rescheduled by the constructor of the Memtable.
+                owner.signalFlushRequired(AbstractAllocatorMemtable.this,
+                                          ColumnFamilyStore.FlushReason.MEMTABLE_PERIOD_EXPIRED);
+            }
```

Comment:
previously, flush was always rescheduled. now, it is rescheduled only if period expires. is it ok?
I thought the problem was only capturing the memtable reference?

Reply:
Yes, the problem is the capture, but I went and made it fully in line with what the older code did, because I think it is preferable to not reschedule if not expired yet, because the task can be scheduled more than once if the period is changed in the metadata.

Reply:
what if the period is not changed? flush is scheduled in the constructor, then is not rescheduled.
don't we loose the flush?

Reply:
The flush task is scheduled to run after the period expires. We get a schedule when the memtable is created, when the period is changed, and when we delay the period if the memtable is clean. In all of these cases, if the current memtable is still the same, the period would have expired.

The problem is tasks scheduled by earlier memtable's creation -- we shouldn't flush new ones because of them.
In the memtable-specific version this was not a problem, because the flush was connected to that memtable.

Reply:
ok, so essentially we'd like to schedule flush on first write, but we don't do it because of perf reasons?

Reply:
I can't talk about what we want, just how it's been set to work, and that's that memtables are set to flush with a given frequency. The action when the table is empty is just an implementation detail, because from the user's point of view rescheduling is equivalent to switching out an empty memtable.

Reply:
right. thanks for explanations!